src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs

```diff
@@ -0,0 +1,1356 @@
+{
+    public void Sub(IEnumerable<int> h, int x)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        if (x == 1)
+        {
+            [|i|].Union([|h|]).ToList();
+            [|i|].Join([|h|], n => n, n => n, (n, m) => n + m).ToList();
+        }
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
+        public async Task TestOneInvocationAfterIfBranch1()
+        {
+            var code = @"
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub(int b, int[] c)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        if (b == 1)
+        {
+            [|i|].SingleOrDefault();
+            c.Max();
+        }
+        else if (b == 3)
+        {
+            [|i|].ToList();
+            c.Min();
+        }
+        else
+        {
+            [|i|].Min();
+            c.Average();
+        }
+
+        [|i|].Sum();
+        c.First();
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
+        public async Task TestOneInvocationAfterIfBranch2()
+        {
+            var code = @"
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub(int b, IEnumerable<int> j)
+    {
+        var i = Enumerable.Range(1, 10).ToArray();
+        i.First();
+        [|j|].SingleOrDefault();
+
+        if (b == 1)
+        {
+            i.First();
+            [|j|].First();
+        }
+        else if (b == 3)
+        {
+            i.First();
+            [|j|].First();
+        }
+        else if (b == 5)
+        {
+            i.First();
+            [|j|].First();
+        }
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
+        public async Task TestTwoInvocationsAfterIfElseBranch()
+        {
+            var code = @"
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub(bool flag, IEnumerable<int> j)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        var k = Enumerable.Range(1, 10).ToArray();
+        if (flag)
+        {
+            [|i|].Single();
+            [|j|].SingleOrDefault();
+            k.Single();
+        }
+        else
+        {
+            [|i|].Sum();
+            [|j|].ToArray();
+            k.Last();
+        }
+
+        [|i|].ToDictionary(x => x);
+        [|j|].First();
+        k.LastOrDefault();
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
+        public async Task TestTwoInvocationsAfterIfBranch()
+        {
+            var code = @"
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+public class Bar
+{
+    public void Sub(bool flag, IEnumerable<int> j)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        var k = Enumerable.Range(1, 10).ToArray();
```

Comment:
What happens if `k` is declared as `IEnumerable<int> k = Enumerable.Range(1, 10).ToArray();`?

Reply:
Yep, that's another case that I am preparing here 
https://github.com/Cosifne/roslyn-analyzers/tree/dev/shech%2FExtendTheAnalysisKey 
Basically what should be track ideally is 
1. An non-deferred invocation operation. 
2. Parameter or local symbol that might not have creation operation.

But that's another big change within the OperationVisitor, and I would like to spilt them. (This PR is targeting a feature branch now)

I can add a test here to remind us this case

Reply:
I add a test covering the case you mentioned